File: apps/ios/Wewed/Services/ShadowHTTPContracts.swift
import Foundation

public enum ShadowAPIError: Error, Equatable, Sendable {
    case invalidEnvironment
    case transportNotConfigured
    case invalidResponse
}

public struct ShadowRepositoryConfiguration: Equatable, Sendable {
    public let baseURL: URL
    public let environment: NativeDataEnvironment

    public init(baseURL: URL, environment: NativeDataEnvironment = .shadow) throws {
        guard environment == .shadow else {
            throw ShadowAPIError.invalidEnvironment
        }
        try NativeEnvironmentGuard.validate(baseURL: baseURL, environment: environment)
        self.baseURL = baseURL
        self.environment = environment
    }
}

public protocol ShadowAPITransportProtocol: Sendable {
    func get(path: String) async throws -> Data
    func post(path: String, body: Data) async throws -> Data
    func patch(path: String, body: Data) async throws -> Data
}

/// Deliberate default for this sprint. It makes the HTTP repository boundary
/// compilable and testable without accidentally adding real networking.
public struct UnconfiguredShadowAPITransport: ShadowAPITransportProtocol {
    public init() {}

    public func get(path: String) async throws -> Data {
        throw ShadowAPIError.transportNotConfigured
    }

    public func post(path: String, body: Data) async throws -> Data {
        throw ShadowAPIError.transportNotConfigured
    }

    public func patch(path: String, body: Data) async throws -> Data {
        throw ShadowAPIError.transportNotConfigured
    }
}

public enum ShadowEndpoint {
    public static let plannerDashboard = "/mobile-shadow/planner/overview"
    public static let plannerBudget = "/mobile-shadow/planner/budget"
    public static let plannerContributions = "/mobile-shadow/planner/contributions"
    public static let plannerVendors = "/mobile-shadow/planner/vendors"
    public static let plannerSeating = "/mobile-shadow/planner/seating"
    public static let plannerTimeline = "/mobile-shadow/planner/timeline"
    public static let weddingContext = "/mobile-shadow/wedding"
    public static let guests = "/mobile-shadow/guests"
    public static let invitation = "/mobile-shadow/invitation"
    public static let pass = "/mobile-shadow/pass"
}

/// Prepared HTTP adapter for the Planner read surface. It cannot perform
/// network traffic until an explicit ShadowAPITransport is injected.
public struct ShadowPlannerHTTPRepository: PlannerDashboardRepositoryProtocol {
    private let transport: ShadowAPITransportProtocol
    private let decoder: JSONDecoder

    public init(
        configuration: ShadowRepositoryConfiguration,
        transport: ShadowAPITransportProtocol = UnconfiguredShadowAPITransport()
    ) {
        _ = configuration
        self.transport = transport
        self.decoder = JSONDecoder()
    }

    public func getDashboard() async throws -> PlannerDashboardSnapshot {
        try decoder.decode(
            PlannerDashboardSnapshot.self,
            from: await transport.get(path: ShadowEndpoint.plannerDashboard)
        )
    }

    public func getBudgetLines() async throws -> [PlannerBudgetLine] {
        try decoder.decode(
            [PlannerBudgetLine].self,
            from: await transport.get(path: ShadowEndpoint.plannerBudget)
        )
    }

    public func getContributions() async throws -> [PlannerContributionRecord] {
        try decoder.decode(
            [PlannerContributionRecord].self,
            from: await transport.get(path: ShadowEndpoint.plannerContributions)
        )
    }

    public func getVendorEngagements() async throws -> [PlannerVendorEngagement] {
        try decoder.decode(
            [PlannerVendorEngagement].self,
            from: await transport.get(path: ShadowEndpoint.plannerVendors)
        )
    }

    public func getSeatingTables() async throws -> [PlannerSeatingTable] {
        try decoder.decode(
            [PlannerSeatingTable].self,
            from: await transport.get(path: ShadowEndpoint.plannerSeating)
        )
    }

    public func getTimelineEntries() async throws -> [PlannerTimelineEntry] {
        try decoder.decode(
            [PlannerTimelineEntry].self,
            from: await transport.get(path: ShadowEndpoint.plannerTimeline)
        )
    }

    public func getDocuments() async throws -> [PlannerDocumentRecord] {
        // The HTTP Shadow contract has no documents endpoint yet. Private Real Shadow
        // reads contracts/vault objects directly from the selected local account graph.
        []
    }
}

File: apps/android/app/src/main/java/pro/wewed/app/services/ShadowHTTPContracts.kt
package pro.wewed.app.services

import pro.wewed.app.models.NativeDataEnvironment
import pro.wewed.app.models.PlannerActivityItem
import pro.wewed.app.models.PlannerAttentionItem
import pro.wewed.app.models.PlannerAttentionSeverity
import pro.wewed.app.models.PlannerBudgetLine
import pro.wewed.app.models.PlannerContributionRecord
import pro.wewed.app.models.PlannerDashboardSnapshot
import pro.wewed.app.models.PlannerModuleSummary
import pro.wewed.app.models.PlannerSeatingTable
import pro.wewed.app.models.PlannerTimelineEntry
import pro.wewed.app.models.PlannerVendorEngagement

sealed class ShadowAPIError(message: String) : IllegalStateException(message) {
    data object InvalidEnvironment : ShadowAPIError("Shadow repository requires the SHADOW environment.")
    data object TransportNotConfigured : ShadowAPIError("Shadow HTTP transport is not configured.")
}

data class ShadowRepositoryConfiguration(
    val baseUrl: String,
    val environment: NativeDataEnvironment = NativeDataEnvironment.SHADOW
) {
    init {
        if (environment != NativeDataEnvironment.SHADOW) {
            throw ShadowAPIError.InvalidEnvironment
        }
        NativeEnvironmentGuard.validate(baseUrl, environment)
    }
}

interface ShadowAPITransport {
    suspend fun get(path: String): String
    suspend fun post(path: String, body: String): String
    suspend fun patch(path: String, body: String): String
}

class UnconfiguredShadowAPITransport : ShadowAPITransport {
    override suspend fun get(path: String): String = throw ShadowAPIError.TransportNotConfigured
    override suspend fun post(path: String, body: String): String = throw ShadowAPIError.TransportNotConfigured
    override suspend fun patch(path: String, body: String): String = throw ShadowAPIError.TransportNotConfigured
}

object ShadowEndpoint {
    const val PLANNER_DASHBOARD = "/mobile-shadow/planner/overview"
    const val PLANNER_BUDGET = "/mobile-shadow/planner/budget"
    const val PLANNER_CONTRIBUTIONS = "/mobile-shadow/planner/contributions"
    const val PLANNER_VENDORS = "/mobile-shadow/planner/vendors"
    const val PLANNER_SEATING = "/mobile-shadow/planner/seating"
    const val PLANNER_TIMELINE = "/mobile-shadow/planner/timeline"
    const val WEDDING_CONTEXT = "/mobile-shadow/wedding"
    const val GUESTS = "/mobile-shadow/guests"
    const val INVITATION = "/mobile-shadow/invitation"
    const val PASS = "/mobile-shadow/pass"
}

/**
 * Contract placeholder for the future shadow Planner HTTP repository.
 *
 * Android deliberately has no JSON/network dependency added in this phase.
 * The interface is fixed now; transport/serialization is added only after the
 * real Shadow backend contract is approved.
 */
interface ShadowPlannerWireContract {
    suspend fun plannerDashboardJson(): String
    suspend fun budgetJson(): String
    suspend fun contributionsJson(): String
    suspend fun vendorsJson(): String
    suspend fun seatingJson(): String
    suspend fun timelineJson(): String
}

class TransportBackedShadowPlannerWireContract(
    private val configuration: ShadowRepositoryConfiguration,
    private val transport: ShadowAPITransport = UnconfiguredShadowAPITransport()
) : ShadowPlannerWireContract {
    init {
        // Keep configuration strongly referenced and validated at construction.
        NativeEnvironmentGuard.validate(configuration.baseUrl, configuration.environment)
    }

    override suspend fun plannerDashboardJson(): String = transport.get(ShadowEndpoint.PLANNER_DASHBOARD)
    override suspend fun budgetJson(): String = transport.get(ShadowEndpoint.PLANNER_BUDGET)
    override suspend fun contributionsJson(): String = transport.get(ShadowEndpoint.PLANNER_CONTRIBUTIONS)
    override suspend fun vendorsJson(): String = transport.get(ShadowEndpoint.PLANNER_VENDORS)
    override suspend fun seatingJson(): String = transport.get(ShadowEndpoint.PLANNER_SEATING)
    override suspend fun timelineJson(): String = transport.get(ShadowEndpoint.PLANNER_TIMELINE)
}
